taskcat/_cfn/stack.py

```diff
@@ -50,59 +63,73 @@ class Capabilities:
 
     def __str__(self):
-        return f"{self.timestamp} {self.logical_id} {self.status}"
+        return "{} {} {}".format(self.timestamp, self.logical_id, self.status)
 
     def __repr__(self):
-        return f"<Event object {self.event_id} at {hex(id(self))}>"
+        return "<Event object {} at {}>".format(self.event_id, hex(id(self)))
 
 
 class Resource:
     def __init__(
-        self,
-        stack_id: str,
-        raw_resource: dict,
-        test_name: str = "",
-        uuid: Optional[UUID] = None,
+        self, stack_id: str, resource_dict: dict, test_name: str = "", uuid: UUID = None
     ):
+        uuid = uuid if uuid else uuid4()
         self.stack_id: str = stack_id
         self.test_name: str = test_name
-        self.uuid: UUID = uuid if uuid else uuid4()
-        self.logical_id: str = raw_resource["LogicalResourceId"]
-        self.type: str = raw_resource["ResourceType"]
-        self.status: str = raw_resource["ResourceStatus"]
-        self.physical_id: str = raw_resource.get("PhysicalResourceId", "")
-        self.last_updated_timestamp: datetime = raw_resource.get(
-            "LastUpdatedTimestamp", datetime.fromtimestamp(0)
-        )
+        self.uuid: UUID = uuid
+        self.logical_id: str = resource_dict["LogicalResourceId"]
+        self.type: str = resource_dict["ResourceType"]
+        self.status: str = resource_dict["ResourceStatus"]
+        self.physical_id: str = ""
+        self.last_updated_timestamp: datetime = datetime.fromtimestamp(0)
+        self.status_reason: str = ""
+        if "PhysicalResourceId" in resource_dict.keys():
+            self.physical_id = resource_dict["PhysicalResourceId"]
+        if "LastUpdatedTimestamp" in resource_dict.keys():
+            self.last_updated_timestamp = resource_dict["LastUpdatedTimestamp"]
+        if "ResourceStatusReason" in resource_dict.keys():
+            self.status_reason = resource_dict["ResourceStatusReason"]
 
     def __str__(self):
-        return f"<Resource {self.logical_id} {self.status}>"
+        return "<Resource {} {}>".format(self.logical_id, self.status)
 
 
 class Parameter:
-    def __init__(self, raw_param: dict):
-        self.key: str = raw_param["ParameterKey"]
-        self.value: str = raw_param.get("ParameterValue", "")
-        self.use_previous_value: bool = raw_param.get("UsePreviousValue", False)
-        self.resolved_value: str = raw_param.get("ResolvedValue", "")
+    def __init__(self, param_dict: dict):
+        self.key: str = param_dict["ParameterKey"]
+        self.value: str = ""
+        self.raw_value: str = ""
+        self.use_previous_value: bool = False
+        self.resolved_value: str = ""
+        if "ParameterValue" in param_dict.keys():
+            self.value = param_dict["ParameterValue"]
+        if "UsePreviousValue" in param_dict.keys():
+            self.use_previous_value = param_dict["UsePreviousValue"]
+        if "ResolvedValue" in param_dict.keys():
```

Comment:
`ResolvedValue` ?
`UsePreviousValue` ?

I don't recall these being available in this form in the v0.8 branch. What's the mechanism for addressing this with in-config params? (inline)

Reply:
these are part of the cfn describe-stack api response, and will only be populated for stacks imported from existing, can consider them read only (ie. you cannot specify them in a config).

They have no practical use today, just including them so as to map the api as closesly as possible, maybe one day we find a use for them.

Reply:
UsePreviousValue is not exactly read-only, it can be specified when doing a stack update